### PR TITLE
Virt-Who Checkin Performance Improvements

### PR DESCRIPTION
**Pull request by dgoodwin**
virt-who checkins are easily blowing past the 60 second timeout that will occur with both virt-who or subscription-manager. Approx 150 hypervisors each with 15 guests is about enough to do it, in my system this was taking around 50s. 

Problem boiled down to looking up the host consumer record for every hypervisor record, and then two more queries to lookup the guest consumer and it's latest host for each guest ID on each hypervisor.

Fixed by batching up the requests and passing them in memory, so we lookup hypervisor consumers for all hypervisors in the checkin, and similarly the guest and host consumers for all guest IDs. These three large queries vs ~4500 small queries results in a close to 10x performance improvement. (50 -> 7s on my system with the sample data mentioned above)

Added a DTO for managing the intricacies of virt UUID casing and endianness when mapping these to their consumer records.

### CODE REVIEW

File: server/src/main/java/org/candlepin/model/ConsumerCurator.java [hunk 1]
```diff
@@ -169,6 +169,68 @@ public Consumer findByVirtUuid(String uuid, String ownerId) {
...
+        List<String> possibleGuestIds = Util.getPossibleUuids(guestIds.toArray(
+                new String [guestIds.size()]));
+
+        String sql = "select cp_consumer.uuid from cp_consumer " +
+            "inner join cp_consumer_facts " +
+            "on cp_consumer.id = cp_consumer_facts.cp_consumer_id " +
+            "where cp_consumer_facts.mapkey = 'virt.uuid' and " +
+            "lower(cp_consumer_facts.element) in (:guestids) " +
+            "and cp_consumer.owner_id = :ownerid " +
+            "order by cp_consumer.updated desc";
```
Comment (mstead) [inline server/src/main/java/org/candlepin/model/ConsumerCurator.java]: Any reason for using pure SQL here? Does this run on oracle?

Comment (dgoodwin) [inline server/src/main/java/org/candlepin/model/ConsumerCurator.java]: Each of the new queries is based on the singular version of the query we were using before (findByVirtUuid right above). Not sure of the original reasons but should be fine on Oracle yes.

Comment (dgoodwin) [inline server/src/main/java/org/candlepin/model/ConsumerCurator.java]: Think they should be rewritten? Pretty sure it was by Carter and I suspect there's a good reason but we could try to find out what it was.

Comment (mstead) [inline server/src/main/java/org/candlepin/model/ConsumerCurator.java]: Lets just leave it if it is working an working well -- lets change it if it becomes an issue.

File: server/src/main/java/org/candlepin/model/ConsumerCurator.java [hunk 2]
```diff
@@ -450,17 +549,42 @@ public Consumer getHypervisor(String hypervisorId, Owner owner) {
...
     @SuppressWarnings("unchecked")
     @Transactional
-    public List<Consumer> getHypervisorsBulk(List<String> hypervisorIds, String ownerKey) {
+    public List<Consumer> getHypervisorsBulk(Collection<String> hypervisorIds,
+            String ownerKey) {
         if (hypervisorIds == null || hypervisorIds.isEmpty()) {
             return new LinkedList<Consumer>();
         }
-        return createSecureCriteria()
+        return currentSession().createCriteria(Consumer.class)
```
Comment (mstead) [inline server/src/main/java/org/candlepin/model/ConsumerCurator.java]: Just curious -- why was this changed from secure.

File: server/src/main/java/org/candlepin/resource/ConsumerResource.java [hunk 3]
```diff
@@ -1026,10 +1044,13 @@ private boolean checkForGuestsUpdate(Consumer existing, Consumer incoming) {
             }
 
             // The guest has not registered. No need to process entitlements.
-            if (guest == null) {
+            if (guestConsumerMap.get(guestId.getGuestId()) == null) {
                 continue;
             }
 
+            Consumer guest = guestConsumerMap.get(guestId.getGuestId());
```
Comment (mstead) [inline server/src/main/java/org/candlepin/resource/ConsumerResource.java]: Could avoid a 2nd map lookup by moving this above the if -- just being picky cuz I can't find anything else. :)

**Comment by mstead** [comment]
@dgoodwin this looks great. Easy to see how this change would boost performance. Also like the mapping Object -- very handy for passing the relationships around. :+1: 

Let me know when you are ready and I'll merge this in.

File: server/src/main/java/org/candlepin/model/ConsumerCurator.java (hunk 2 above)
Comment (dgoodwin) [inline server/src/main/java/org/candlepin/model/ConsumerCurator.java]: Because a hypervisor checkin is done as a consumer, they typically cannot see other consumers, so with secure criteria the automatic filtering kicks in and hides all the host consumers. This is why the singular getHost does unsecured as well. This actually tripped me up for hours trying to figure out why it wasn't seeing the hosts and trying to re-create them. :)

File: server/src/main/java/org/candlepin/resource/ConsumerResource.java (hunk 3 above)
Comment (dgoodwin) [inline server/src/main/java/org/candlepin/resource/ConsumerResource.java]: Addressed! :)
